numpy/core/function_base.py

```diff
@@ -83,28 +83,45 @@ def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None):
     """
     num = int(num)
 
-    # Convert float/complex array scalars to float, gh-3504 
+    # Convert float/complex array scalars to float, gh-3504
     start = start * 1.
     stop = stop * 1.
 
+    dt = result_type(start, stop, float(num))
     if dtype is None:
-        dtype = result_type(start, stop, float(num))
+        dtype = dt
 
     if num <= 0:
         return array([], dtype)
+    if num == 1:
+        return array([start], dtype=dtype)
+    y = _nx.arange(0, num, dtype=dt)
+    if endpoint:
+        num -= 1
+    y *= 1/num
```

Comment:
There's a `from __future__ import division` at the top of the file, so all divisions are true divisions.